Add queue monitoring

fearful-symmetry opened the pull request:
This is part of: https://github.com/elastic/beats/issues/31113 and https://github.com/elastic/elastic-agent-shipper/issues/11

This is a very basic but functional metrics reporter that fetches queue metrics from the queue's `Metrics()` function and sends it to a user-configured output. In its current state, it has two outputs: a logger, similar to the one in metricbeat, and an `expvar` output that sends the queue metrics to `/vars/debug`. It can integrate with the rest of the config, and there's updated example code in `elastic-agent-shipper.yml`

Right now we have no output connected, as we're still waiting on https://github.com/elastic/beats/pull/31356 and other improvements before the memory queue ends up in here. As such, there's a bit of awkwardness in this code, particularly in where it connects to the shipper, as I'm not quite sure what things will look like once the shipper actually does something. I get the feeling that most of the code in `server/run.go` is going to change once we have a queue and output set up, but for now it's fairly threadbare.

I ran into a ton of footcannons and issues with the `config` library while working on this, and it makes me wonder if that library, which is a tad old and was built with the assumptions of beats in mind, is a good fit for something like the shipper. The library is very unopinionated and vague, which can be a problem both for us (lots of ugliness to mash it into a concrete type) and the users (a small misconfiguration can result in some subtle no-ops). However, this is outside the scope of this PR.

Not setting any reviewers right now, as I assume something in CI is instantly gonna break.

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-mbp%2FPR-29/detail/PR-29/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-mbp%2FPR-29/detail/PR-29/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-mbp%2FPR-29/detail/PR-29/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-mbp%2FPR-29/detail/PR-29/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent-shipper_29.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2022-04-21T17:47:11.285Z&rangeTo=2022-04-21T18:07:11.285Z&transactionName=elastic-agent-shipper-mbp/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=6da7acee7b4c4b8aada6d633aca9b56c&transactionId=88c6f27c91fcfba6)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-04-21T17:57:11.285+0000


* Duration: 19 min 15 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 2  |
| Passed       | 0  |
| Skipped      | 0 |
| Total        | 2   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![2](https://img.shields.io/badge/2%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-mbp%2FPR-29/detail/PR-29/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Test / Matrix - PLATFORM = 'ubuntu-20.04 && immutable' / Test / TestSetupMonitor – github.com/elastic/elastic-agent-shipper/monitoring`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestSetupMonitor
panic: test timed out after 10m0s

goroutine 3 [running]:
testing.(*M).startAlarm.func1()
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1788 +0x8e
created by time.goFunc
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/time/sleep.go:180 +0x31

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00022f520, {0xb97b9c, 0x46c6d3}, 0xbc1798)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1307 +0x375
testing.runTests.func1(0xc000259830)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1598 +0x6e
testing.tRunner(0xc00022f520, 0xc0002efd18)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1259 +0x102
testing.runTests(0xc000174e80, {0x117da20, 0x3, 0x3}, {0x486f8d, 0xb96b7e, 0x118d760})
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1596 +0x43f
testing.(*M).Run(0xc000174e80)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1504 +0x51d
main.main()
	_testmain.go:47 +0x14b

goroutine 21 [chan send (nil chan)]:
github.com/elastic/elastic-agent-shipper/monitoring.QueueMonitor.End(...)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/src/github.com/elastic/elastic-agent-shipper/monitoring/queuemon.go:99
github.com/elastic/elastic-agent-shipper/monitoring.TestSetupMonitor(0x0)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/src/github.com/elastic/elastic-agent-shipper/monitoring/queuemon_test.go:118 +0x1f0
testing.tRunner(0xc00022f6c0, 0xbc1798)
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/var/lib/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.linux.amd64/src/testing/testing.go:1306 +0x35a
 
 ```
</p></details>
</ul>
    
##### `Test / Matrix - PLATFORM = 'windows-2019 && windows-immutable' / Test / TestSetupMonitor – github.com/elastic/elastic-agent-shipper/monitoring`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestSetupMonitor
panic: test timed out after 10m0s

goroutine 18 [running]:
testing.(*M).startAlarm.func1()
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1788 +0x8e
created by time.goFunc
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/time/sleep.go:180 +0x31

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00017f520, {0x19605a9, 0x126da33}, 0x198aa80)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1307 +0x375
testing.runTests.func1(0xc000234990)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1598 +0x6e
testing.tRunner(0xc00017f520, 0xc0002dfd18)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1259 +0x102
testing.runTests(0xc00012c980, {0x1f315a0, 0x3, 0x3}, {0x129440d, 0x195f2fa, 0x1f40e00})
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1596 +0x43f
testing.(*M).Run(0xc00012c980)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1504 +0x51d
main.main()
	_testmain.go:47 +0x14b

goroutine 6 [chan send (nil chan)]:
github.com/elastic/elastic-agent-shipper/monitoring.QueueMonitor.End(...)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/src/github.com/elastic/elastic-agent-shipper/monitoring/queuemon.go:99
github.com/elastic/elastic-agent-shipper/monitoring.TestSetupMonitor(0x0)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/src/github.com/elastic/elastic-agent-shipper/monitoring/queuemon_test.go:118 +0x1f0
testing.tRunner(0xc00017f6c0, 0x198aa80)
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	C:/Users/jenkins/workspace/elastic-agent-shipper-mbp_PR-29/.gvm/versions/go1.17.8.windows.amd64/src/testing/testing.go:1306 +0x35a
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-mbp%2FPR-29/detail/PR-29/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Running Go tests`
<ul>
<li>Took 10 min 12 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent-shipper-mbp/pipelines/PR-29/runs/1/steps/234/log/?start=0">here</a></li>
<li>Description: <code>gotestsum --format testname --junitfile junit-report.xml -- -v ./...</code></l1>
</ul>
  
  
  
  
##### `Running Go tests`
<ul>
<li>Took 10 min 15 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent-shipper-mbp/pipelines/PR-29/runs/1/steps/313/log/?start=0">here</a></li>
<li>Description: <code>gotestsum --format testname --junitfile junit-report.xml -- -v ./...</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

cmacknz commented on elastic-agent-shipper.yml:
```diff
@@ -7,4 +7,16 @@ tls: false
+queue:
+  test: #There is no actual "test" queue type, remove this later.
+    events: 512
+  metrics:
```
I'm not sure we want this to be a property of the queue. It should probably be a top level level configuration.

Also consider that users are unlikely to be configuring the shipper, they will be configuration the agent (likely with the output section being translated here).

We want the configuration applied to the shipper to come directly from the agent policy without any transformation. How does the agent enable the HTTP expvar/metrics endpoint? Does it allow that? Should this instead be collected using the diagnostic command interface and the command protocol.

I realize this is for initial development but we shouldn't let this diverge too much from what the final system should do.

fearful-symmetry replied:
I was actually really on the fence about where the config should "live" but that makes sense, yah.

cmacknz replied:
Does the agent use the same configuration library as beats does? We may want to use the same configuration package the agent does if it doesn't to minimize any possibility for incompatibility.

faec commented on config/config.go:
```diff
@@ -33,11 +33,12 @@ func init() {
+	Cert  string      `config:"cert"`  //TLS cert file, if TLS is enabled
+	Key   string      `config:"key"`   //TLS Keyfile, if specified
+	Port  int         `config:"port"`  //Port to listen on
+	Queue *config.C   `config:"queue"` //Shipper queue Settings
```
Can this be an explicit type rather than using `ucfg`? Just making a `QueueConfig` struct with the fields you need for now is fine (looks like maybe just `Metrics`?), other config fields can be added as needed. I just want to make sure we keep things in explicit structs as much as possible so we don't lose track of what fields are in use the way we did in beats.

fearful-symmetry replied:
Yah, I put this in as a placeholder while I was working with the dummy queue, we might want to remove it entirely until we have an actual queue hooked up.

faec commented on config/config.go:
```diff
@@ -33,11 +33,12 @@ func init() {
+	Cert  string      `config:"cert"`  //TLS cert file, if TLS is enabled
+	Key   string      `config:"key"`   //TLS Keyfile, if specified
+	Port  int         `config:"port"`  //Port to listen on
+	Queue *config.C   `config:"queue"` //Shipper queue Settings
```
Can this be an explicit type rather than using `ucfg`? Just making a `QueueConfig` struct with the fields you need for now is fine (looks like maybe just `Metrics`?), other config fields can be added as needed. I just want to make sure we keep things in explicit structs as much as possible so we don't lose track of what fields are in use the way we did in beats.

faec commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,208 @@
+
+//Config is the intermediate struct representation of the queue monitor config
+type Config struct {
+	OutputConfig []config.Namespace `config:"outputs"`
```
Same as with the queue config, I'd rather this was an explicit empty/placeholder type than an untyped key-value tree. For this pass we're just aiming for a bare test output anyway, so there isn't yet any configuration to propagate, and once there is I'd like to keep it as explicitly-typed as we can. (Is this field used for anything currently? If not it could just be removed for now.)

faec replied:
Same as with the queue config, I'd rather this was an explicit empty/placeholder type than an untyped key-value tree. For this pass we're just aiming for a bare test output anyway, so there isn't yet any configuration to propagate, and once there is I'd like to keep it as explicitly-typed as we can. (Is this field used for anything currently? If not it could just be removed for now.)

faec commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,184 @@
+	return TestMetricsReporter{mon: make(chan reporter.QueueMetrics, 10)}, nil
+}
+
+func (tr TestMetricsReporter) Update(data reporter.QueueMetrics) error {
```
`data` -> `metrics`

faec replied:
`data` -> `metrics`

faec reviewed (changes requested):
I agree about the config library, and a lot of my first-pass review is about mitigating that.

The biggest antipattern I've seen come out of this library in beats is to unpack a configuration as different types at different times, with no systematic accounting for what fields might be read from it over the course of the pipeline. This means there's no way to generate an authoritative list of what config fields are valid.

I would really like it if any new components can generate authoritative lists of what config options are valid :-) Since replacing the library dependency is out of scope right now, my suggestion to keep things manageable is to make sure any particular config source has an explicit documented go type, and it's only allowed to be decoded as that. For this PR that mostly means removing the untyped key-value config fields and replacing them with concrete types where we know all the fields, even if they're incomplete for now.

fearful-symmetry commented on monitoring/reporter/reporter.go:
```diff
@@ -0,0 +1,31 @@
+
+// Reporter is the bare interface that will be implemented by the various outputs
+type Reporter interface {
+	Update(QueueMetrics) error
```
I'm honestly not sure how the usage of this will expand over time, so I tried to be as general as possible with a lot of the interfaces. However, for the sake of being extensible I'm leaning towards the latter.

fearful-symmetry commented on server/example_queue.go:
```diff
@@ -0,0 +1,57 @@
+
+// ===================================== Test code, remove this later
+
+// TestMetricsQueue is a test queue for the reporter
```
Yah, committed this on accident, oop

fearful-symmetry commented on monitoring/reporter/register.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+//The register defines a global interface for managing the various metrics outputs
+//This might be overkill for what we have now, but I'd like to at least try to prepare
```
Agreed that it would be preferable to not use the `config.Namespace` type. Not sure what you mean by `For this phase we're specifically scoped to a single output per executable` ? Do you mean a single queue monitoring output? I was kinda going off this assumption that the logging monitoring and `expvar` weren't mutually exclusive, but more "default always on" and "optional debugging interface"

cmacknz commented on elastic-agent-shipper.yml:
```diff
@@ -7,4 +7,19 @@ tls: false
+  test: #There is no actual "test" queue type, remove this later.
+    events: 512
+
+metrics:
```
You may want to have this match the agent monitoring configuration, then it could just pass it through:

```yaml
agent.monitoring:
  # enabled turns on monitoring of running processes
  enabled: false
  # enables log monitoring
  logs: false
  # enables metrics monitoring
  metrics: false
  # exposes agent metrics using http, by default sockets and named pipes are used
  http:
      # enables http endpoint
      enabled: false
      # The HTTP endpoint will bind to this hostname, IP address, unix socket or named pipe.
      # When using IP addresses, it is recommended to only use localhost.
      host: localhost
      # Port on which the HTTP endpoint will bind. Default is 0 meaning feature is disabled.
      port: 6791
```

https://www.elastic.co/guide/en/fleet/current/elastic-agent-monitoring-configuration.html

cmacknz replied:
If we want to use configuration options that are not already available in the agent configuration, then we need to consider how they get passed down to us. Ideally the shipper is an implementation detail that users do not need to explicitly configure. 

It does probably make sense to have some additional settings we can control the default values of via the shipper binary's agent specification file or something though. Those should either extend the agent monitoring configuration in a way that requires no translation, or just be a separate configuration block.

fearful-symmetry replied:
including the `agent.` prefix?

cmacknz replied:
This is fine, exactly how this should work will be naturally discovered when we integrate with the agent properly.

cmacknz commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	if queueIsFull {
+		mon.queueLimitCount = mon.queueLimitCount + 1
```
Reading this, what information is this providing us with that `QueueIsCurrentlyFull` doesn't give us?

I think where we would this is in the shipper Publish RPC handler, so we could track how many incoming events were rejected because the queue is full. I'm not sure how useful it is here.

cmacknz commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,204 @@
+		//All we have now is total queue size in bytes, which doesn't tell us how many more events could fit before we hit the queue limit.
+		//So until we have something better, mark anything as 90% or more full as "full"
+
+		// I'm assuming that limit can be zero here, as perhaps a user can configure a queue without a limit, and it gets passed down to us.
```
I'm wondering if it is misleading to track this here. It seems like the queue should just report its current fill level, and we should track the count of events that were rejected because they could not fit on the publisher side (likely the RPC handlers).

fearful-symmetry replied:
@faec what do you think? Does it make more sense to track queue levels here, or in the shipper side? My thinking was to keep the shipper `Metrics()` API more low-level, but I'm kinda on the fence.

faec replied:
Well, I do like the idea of reporting the explicit levels without trying to summarize into "full" or "not full" since that's always going to be somewhat brittle.

I'm not sure what the right alternative would be, though. For example, for a queue configured to block rather than rejecting when full, the important question might be how many shipper requests blocked for a nontrivial duration, or the total amount of time consumers spent waiting for queue space over the last [time interval].

I'm not sure we know enough to pick the right summary fields yet, though. I expect we'll have more basis for deciding as we develop standard benchmarks. So from that perspective, I'd be fine keeping a "full" field that just checks a 90% threshold, and leaving a comment saying that this is a placeholder to show off the monitoring framework and should be replaced by a more principled summary once we've collected enough data.

fearful-symmetry replied:
Yah, agreed that a lot of these summary fields are definitely gonna change over time. I'm thinking we should be comfortable with expanding/tweaking these metrics over the next few releases. A lot of these metrics might make more/less sense once we have the actual queues and everything hooked up too.

cmacknz replied:
Makes sense to me. The obvious place to measure this will likely show up as we build more of the shipper.

cmacknz commented on monitoring/reporter/expvar/expvars.go:
```diff
@@ -0,0 +1,70 @@
+
+func (exp Expvars) runFrontend(cfg ExpvarsConfig) {
+	go func() {
+		err := http.ListenAndServe(cfg.Addr, nil)
```
It looks like we need a way to trigger the HTTP server to stop and force the goroutine to exit.

fearful-symmetry replied:
Agreed, starting to think the sigint handling stuff should just be crammed into this PR.

faec commented on monitoring/reporter/expvar/expvars.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+//ExpvarsConfig is the config struct for marshalling whatever we get from the config file
+type ExpvarsConfig struct {
```
This should probably be `Config`, for consistency

faec reviewed (approved):
Thanks! I'm pretty happy with it at this point. The one "big" suggestion I would make is changing the language of `output` (variable / function names, config fields) to `reporter` instead. This is subjective, but I'd rather avoid semantic collision with the "output" of the shipper itself (which is what I initially thought the registry was preparing for), and "reporter" keeps a clear distinction between data throughput and monitoring frontends.

That is a soft request though, I'm open to alternatives. Otherwise I think I'm happy with it, so it would be nice if others can still take a look but I'm approving pending whatever naming changes are decided on.

cmacknz commented:
> This is subjective, but I'd rather avoid semantic collision with the "output" of the shipper itself (which is what I initially thought the registry was preparing for), and "reporter" keeps a clear distinction between data throughput and monitoring frontend

I agree with this now that @faec has pointed it out. Let's avoid overloading the word output even though it is technically correct as a description here.

fearful-symmetry commented:
Agreed on changing the output names.

fearful-symmetry commented:
Alright, changed a bunch of output names, added the signal handling code. Still waiting for further discussion on some of the bigger "where is this stuff calculated" questions.

cmacknz commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,207 @@
+
+//Config is the intermediate struct representation of the queue monitor config
+type Config struct {
+	Outputs  ReporterConfig `config:"outputs"`
```
```suggestion
	Reporters  ReporterConfig `config:"outputs"`
```

There are a few places where output is still used as the name.

cmacknz commented on monitoring/reporter/expvar/expvars.go:
```diff
@@ -0,0 +1,90 @@
+		err := srv.ListenAndServe()
+		if err != nil {
+			// Error type isn't happy with %w here
+			exp.log.Errorf("Error starting HTTP expvar server: %s", err)
```
You may want to filter out `http.ErrServerClosed` here, which will always happen on shutdown.

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,190 @@
+	mon, err := NewFromConfig(monitor, queue)
+	assert.NoError(t, err)
+	mon.Watch()
+	time.Sleep(time.Second)
```
Why is this delay necessary? Can we get rid of it?

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,190 @@
+
+	t.Logf("listening for events...")
+	// once we have maxEvents, we can properly
+
```
```suggestion
```

Extra space can be removed

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,190 @@
+			break
+		}
+
+		time.Sleep(time.Second)
```
You can probably configure this go faster than once per second. One second is forever to a computer :P

fearful-symmetry commented:
Okay, so I've changed the monitoring config to be a little more "agent-like" in terms of the keys. I'm still on the fence about actually calling the config `agent.monitoring` as this is technically a separate library that doesn't accept 100% of the same values, and mirroring the namespaces seems like a recipe for confusion.

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,186 @@
+		Enabled:  true,
+		ExpvarOutput: expvar.Config{
+			Enabled: true,
+			Port:    8081,
```
I think this is always starting a server listening on localhost port 8081. What you probably want to do instead if listen on a randomly chosen unallocated port (port 0 can request this). That will prevent port collisions in tests.

You can use https://pkg.go.dev/net/http/httptest#NewServer to wrap this up in a new way but you'd have to add a way to change what method is used to start the server.

fearful-symmetry replied:
Yah, agreed, that might be what is nuking the tests, since it works on my system.

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,186 @@
+			Name:    name,
+		},
+	}
+
```
```suggestion
```
nit: bonus whitespace

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,186 @@
+	assert.NoError(t, err)
+	mon.Watch()
+	mon.End()
+
```
```suggestion
```
nit: bonus whitespace

cmacknz reviewed (approved):
Changes LGTM once you fix the test failures.

rdner commented on elastic-agent-shipper.yml:
```diff
@@ -7,4 +7,18 @@ tls: false
+    enabled: true
+    host: "localhost"
+    port: 8282
+    name: "queue"
```
It's unclear to me what's the purpose of this file, is it used for testing? A short heading comment would help here :)

fearful-symmetry replied:
This is the example config. Added a little comment.

AndersonQ commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,186 @@
+	var maxEvents uint64 = 10
+	queue := NewTestQueue(maxEvents)
+	mon, err := NewFromConfig(monitor, queue)
+	assert.NoError(t, err)
```
[Suggestion]
Use [`require.NoError`](https://pkg.go.dev/github.com/stretchr/testify/require#NoError), it'll abort the test if check fails. This is better for checking required conditions.

AndersonQ commented on monitoring/queuemon_test.go:
```diff
@@ -0,0 +1,186 @@
+
+	}
+	mon.End()
+	assert.True(t, gotQueueIsFull, "Did not report a full queue")
```
How could that not be true if when you exit the loop, you set it to true?
This test is only testing the other 2 values, `limitCount` and `queueFullCount`.

AndersonQ commented on monitoring/reporter/expvar/expvars.go:
```diff
@@ -0,0 +1,92 @@
+	srv     *http.Server
+}
+
+// NewExpvarReporter initializes the expvar interface, and starts the http frontend.
```
[Nit | suggestion]
from my perspective a concrete type, that implements an interface is initializes, not the interface itself as the test here suggests.
```suggestion
// NewExpvarReporter initializes the expvar reporter, and starts the http frontend.
```

AndersonQ commented on monitoring/reporter/expvar/expvars.go:
```diff
@@ -0,0 +1,92 @@
+	}()
+}
+
+func (exp *Expvars) format() interface{} {
+	to := mapstr.M{}
+
+	err := typeconv.Convert(&to, exp.metrics)
+	// This function is a callback sent to expvar.Func, so this is probably the best way to handle errors.
```
By _this function_ you mean the `func (exp *Expvars) format()` method? If so I'd suggest to move the comment to the beginning of the method. 
```suggestion
// format is a callback sent to expvar.Func, so this is probably the best way to handle errors.
func (exp *Expvars) format() interface{} {
	to := mapstr.M{}
	err := typeconv.Convert(&to, exp.metrics)
```

AndersonQ commented on monitoring/reporter/reporter.go:
```diff
@@ -0,0 +1,32 @@
+type QueueMetrics struct {
+	//QueueLimitReachedCount is the number of times the queue has reached its user-configured limit.
+	QueueLimitReachedCount opt.Uint `struct:"queue_limit_reached_count,omitempty" json:"queue_limit_reached_count"`
+	//QueueIsCurrentlyFull reports if the queue is currently at its user-configured limit
+	QueueIsCurrentlyFull bool `struct:"queue_is_currently_full,omitempty" json:"queue_is_currently_full"`
+	//CurrentQueueLevel reports the current fill state of the queue, in the native user-configured limits of the queue
+	CurrentQueueLevel opt.Uint `struct:"current_queue_level,omitempty" json:"current_queue_level"`
```
[Question | Suggestion]
It's my understanding that these metrics are already the current state of the queue, therefore I find it a bit redundant to have _current/currently_ for only these 2 metrics. Besides the other metrics do not have it. What could be understood as they being not the current state.

```suggestion
	// Full reports if the queue is currently at its user-configured limit
	 Full bool `struct:"queue_is_currently_full,omitempty" json:"queue_is_currently_full"`
	// Level reports the current fill state of the queue, in the native user-configured limits of the queue
	Level opt.Uint `struct:"current_queue_level,omitempty" json:"current_queue_level"`
```

AndersonQ commented on server/run.go:
```diff
@@ -34,6 +39,26 @@ func LoadAndRun() error {
+	go func() {
+		sig := <-sigc
+
+		switch sig {
+		case syscall.SIGINT, syscall.SIGTERM:
+			log.Debug("Received sigterm/sigint, stopping")
+		case syscall.SIGHUP:
+			log.Debug("Received sighup, stopping")
+		}
```
+1 nice :)

AndersonQ commented on monitoring/reporter/reporter.go:
```diff
@@ -0,0 +1,32 @@
+	"github.com/elastic/elastic-agent-libs/opt"
+)
+
+// The reporter interface allows for the creation of new metrics outputs that can sent metrics to any user-configured output. Expvar, logs, etc
```
[Suggestion]
This comment is about the `Reporter` interface, and its definition is quite far away from this comment. The comment feels, to me, lost here, as commenting something that was removed. I'd suggest to put it as part of the `Reporter` doc string

AndersonQ reviewed (approved):
The code is quite well commented, really nice job! :)

I did not find any blocker, I added a few questions and comments. I see there seem to be a mix of concepts or "leaking" abstraction regarding a _reporter_ and an _output_. The code uses the _reporter_ but your comments refer several times to an _output_. I'd suggest to stick with reporter.

Also there are a few linting issues and inconsistency I can spot. I commented in some of them, but in general there is a lack of consistency in the comments some follow Go's conventions, others not. I don't know the conventions on the repo, but I'd suggest to stick with the Go convention.

Also the package is already monitoring, I see as necessary to suffix the `queue` files with `mon`. I'd rather use `monitoring/queue.go` and `monitoring/queue_test.go`. The package name already makes it clear it's about monitoring.

Anyway, these are my suggestions, none is a blocker, neither I'm familiar with the repo's conventions, so fell free to ignore them.

fearful-symmetry commented on elastic-agent-shipper.yml:
```diff
@@ -7,4 +7,18 @@ tls: false
+logging.to_stderr: true
+
+queue:
+  test: #There is no actual "test" queue type, remove this later.
+    events: 512
```
Whenever we get an actual queue hooked up to this, which will be sometime later, and is happening in parallel to this.

fearful-symmetry commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,198 @@
+	if mon.bypass {
+		return
+	}
+	mon.log.Infof("Shutting down metrics monitor...")
+	for _, out := range mon.reporters {
+		out.Close()
+	}
```
The closers have hard timeouts, so I'm not anticipating this being an issue, at least for now?

fearful-symmetry commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,198 @@
+func (mon QueueMonitor) sendToReporters(metrics reporter.QueueMetrics) {
+	for _, out := range mon.reporters {
+		err := out.ReportQueueMetrics(metrics)
+		//Assuming we don't want to make this a hard error, since one broken output doesn't mean they're all broken.
+		if err != nil {
+			mon.log.Errorf("Error sending to output: %w", err)
```
In theory the reporters could fail due to some user misconfiguration or system issue, so I'd rather put a big flashing light in front of them, so to speak.

fearful-symmetry commented on monitoring/queuemon.go:
```diff
@@ -0,0 +1,198 @@
+		//All we have now is total queue size in bytes, which doesn't tell us how many more events could fit before we hit the queue limit.
+		//So until we have something better, mark anything as 90% or more full as "full"
+
+		// I'm assuming that limit can be zero here, as perhaps a user can configure a queue without a limit, and it gets passed down to us.
```
A lot of the underlying `Metrics()` API itself is currently in flux, so I'm kind of assuming a bit of this is going to change over time.

cmacknz commented on monitoring/queuemon_test.go:
```diff
@@ -81,21 +85,22 @@ func (tq *TestMetricsQueue) Metrics() (queue.Metrics, error) {
 
-// ============ mocked output reporter
+func getRandPort() int {
+	return rand.Intn(65535-49152) + 49152 //nolint:gosec //This is a test, strong crypto not needed
```
Any reason you can't just bind to port 0? That is exactly what httptest does: https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/net/http/httptest/server.go;l=68;drc=refs%2Ftags%2Fgo1.18.1

If you want the allocated port value you'd have to get it out of the net.Listener (again httptest does this in a nicer way): https://stackoverflow.com/a/43425461

This doesn't guarantee you get a free port, just a random port.

fearful-symmetry replied:
Ah, didn't even think of that.
